build-system/eslint-rules/html-template.js

```diff
@@ -159,32 +147,146 @@ module.exports = function (context) {
+    context.report({
+      node,
+      message: `Do not implicitly return ${node.tag.callee.name}()\`...\`, since it requires a helper constant in function scope.`,
+      fix: function* (fixer) {
+        yield fixer.insertTextBefore(node, `{\nreturn (`);
+        yield fixer.insertTextAfter(node, `);\n}`);
+      },
+    });
+  }
 
-      const {callee} = node;
-      if (callee.type !== 'Identifier') {
-        return;
-      }
+  const checkedScopeBlocks = new Set();
+  function tagMustBeHelperConstant(node) {
+    const {name} = node.tag.callee;
+    // Fixer inserts constant when required:
+    //   const html = htmlFor(...);
+    const helperConstName = staticTemplateFactories[name];
+    context.report({
+      node,
+      message: `Do not use ${name}() call directly as template tag. Store its result in a helper constant named "${helperConstName}", and use that as the template tag. This enforces code format, and enables syntax highlighting in some editors.`,
+      fix: function* (fixer) {
+        let scope = context.getScope();
+        // Find the highest scope within the same function block that calls this
+        // factory function, so that the helper is shared.
+        while (
+          scope.type !== 'function' &&
+          scope.upper &&
+          scope.upper.references.filter(
+            ({identifier}) => identifier && identifier.name === name
+          ).length > 0
+        ) {
+          scope = scope.upper;
+        }
+        // A scope's block may not take children directly, we need to insert
+        // it as part of its body field.
+        let insertionBlock = scope.block;
+        insertionBlock =
+          insertionBlock.body && !Array.isArray(insertionBlock.body)
+            ? insertionBlock.body
+            : insertionBlock;
+        if (!checkedScopeBlocks.has(insertionBlock)) {
+          checkedScopeBlocks.add(insertionBlock);
+          if (!scope.set.get(helperConstName)) {
+            // Insert const as close as possible above the node, while also
+            // being a direct child of the selected scope's block.
+            let insertBeforeNode = node;
+            while (insertBeforeNode.parent !== insertionBlock) {
+              insertBeforeNode = insertBeforeNode.parent;
+            }
+            const helperConstInit = context.getSourceCode().getText(node.tag);
+            yield fixer.insertTextBefore(
+              insertBeforeNode,
+              `const ${helperConstName} = ${helperConstInit};\n`
+            );
+          }
+        }
 
-      if (staticTemplateTags.has(callee.name)) {
-        return tagCannotBeCalled(node);
-      }
-      if (staticTemplateFactoryFns.has(callee.name)) {
-        return factoryUsage(node);
+        // htmlFor(...)`...` to html`...`
+        yield fixer.replaceText(node.tag, helperConstName);
+
+        // Prettier formats this in a silly way unless the quasi's content
+        // starts and ends with newlines.
```

Comment:
Curious if changing the [`htmlWhitespaceSensitivity`](https://prettier.io/docs/en/options.html#html-whitespace-sensitivity) Prettier option will help here.

Reply:
And if it helps, there's also an [`lwc`](https://prettier.io/docs/en/options.html#parser) parser that can be used as an alternative to the default `html` parser. Looks like it treats template attributes slightly differently.

Reply:
[It does help.](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAKgCwJYDOGAZhAE4C2AhjEcXlHEgDpQ4wUA2ABhgDwBXTgD4+nPMOhw+qcaLkwA7hBlyZQ4Rm4BuVq3QYAInAAOCACYMA5hmgZ2XAOr54BE1TBwAyggJ4YeABu-gCetiYB0AA0GDD4dOTUtCxQBgC0GMzgBARZSBgARlTmnGEAFOZwxFRCMACU+pgZWQQwZHhgMHn25O0FpY0YzSB4VlDkcN30jKwOPPwaYhJSqhJLwkoqsmuoGjqsIFEgEBF40ATIoFRkZBCKAArXvsggVJyKVCEXRwVkHgDWcBgXncYGsyDaAjgRzgFAKcHMlXMABkqFArAIqFY4AAxRI0ALol41GAQQ4gObOfxwNweby+fxBUIvUbjMhwckMAhwMgwe5-KzUZDVTjco4AKwIAA8AEJ-MCA4FUChwZEMDlIEVikCSqVeaycOAARQEEHgwre2vcZG5ZBeRXhnHJJnasEceHMcWQAA4AAxHF0QbmOP4mF4umk8wIco4AR1N8H5JwuKCoBDSjARCPJ7PjeHZ-KxQs1luhIG5FDwELIUKOfnRhpNZo1WrLMCoBXdnpwyAATEc2lQ8OJ0QBhCAUYsgGkAVnJAm5WA7KdbR0CUIAklBKrAvGB2hEAILbrwwEKGi2iuAAX2vQA)

However, [the setting will affect nodes inside the tree](https://prettier.io/blog/2018/11/07/1.15.0.html#whitespace-sensitive-formatting), not just around the root like the rule. This could be fine, but could affect files beyond the current scope in unexpected ways. My opinion is that we should try to enable it separately, then remove this case.

Reply:
Had not seen your comment about the `lwc` parser. It looks to me like we can set the parser for `.html` files, but I'm not aware of how one configures the `parser` for <code>html\`...\`</code> snippets in Javascript.

Reply:
Looks like the `lwc` parser is for https://lwc.dev and does not apply to this use case. In any case, I was merely offering suggestions I knew very little about, and am fully supportive of punting on this 😃

Reply:
Thanks, I wasn't aware of these options.